### PR TITLE
registers_support

### PR DESCRIPTION
**Pull request by kmateuszssak**

### CODE REVIEW

File: backends/ubpf/examples/rate-limiter-structs.p4 [hunk 1]
```diff
@@ -0,0 +1,59 @@
+#include "ubpf_filter_model.p4"
+#include <core.p4>
+
+#include "header.p4"
+#include "parser.p4"
```
Comment (osinstom) [inline backends/ubpf/examples/rate-limiter-structs.p4]: Are the parser.p4 and header.p4 located in repository? We should place them here or move the implementation of header and parsers from those files to this P4 program.

File: backends/ubpf/target.cpp [hunk 2]
```diff
@@ -38,7 +38,39 @@ namespace UBPF {
...
+                                     cstring tblName,
+                                     cstring key) const {
+        builder->appendFormat("ubpf_map_lookup(&%s, ",
+                              tblName.c_str());
+        builder->append("&");
+        builder->append(key);
+        builder->append(")");
+    }
+
+    void UbpfTarget::emitTableUpdate(EBPF::CodeGenInspector *codeGen,
```
Comment (osinstom) [inline backends/ubpf/target.cpp]: Do we need two implementation of `emitTableUpdate` and `emitTableLookup`?

File: backends/ubpf/ubpfControl.cpp [hunk 3]
```diff
@@ -401,42 +568,201 @@ namespace UBPF {
...
+            builder->endOfStatement(true);
+            builder->blockEnd(true);
+            index++;
+        }
+
         builder->newline();
+        builder->blockEnd(true);
+    }
+
+    const IR::Statement *UBPFControl::recursion(const IR::Statement *statement,
```
Comment (osinstom) [inline backends/ubpf/ubpfControl.cpp]: Perhaps we can find a better name than "recursion"? To be more informative.

File: backends/ubpf/ubpfControl.cpp [hunk 4]
```diff
@@ -446,17 +772,40 @@ namespace UBPF {
...
 
-    void UBPFControl::emitTableTypes(EBPF::CodeBuilder* builder) {
+    bool
+    UBPFControl::vairiableIsUsedAsPointer(const IR::Declaration_Variable *vd,
+                                          const IR::AssignmentStatement *assiStat) const {
+        bool pointerType = false;
+        if (assiStat->right->is<IR::MethodCallExpression>()) {
+            auto methCall = assiStat->right->to<IR::MethodCallExpression>();
+            if (methCall->method->is<IR::Member>()) {
+                if (methCall->method->to<IR::Member>()->member.name ==
```
Comment (osinstom) [inline backends/ubpf/ubpfControl.cpp]: Can we squash it to one `if` block?

```
if (methCall->method->is<IR::Member>() && methCall->method->to<IR::Member>()->member.name == UBPFModel::instance.registerModel.read.name) {
```

File: backends/ubpf/ubpfControl.h [hunk 5]
```diff
@@ -26,12 +26,12 @@ namespace UBPF {
 
     class UBPFControlBodyTranslator : public EBPF::CodeGenInspector {
     public:
-        const UBPFControl *control;
+        UBPFControl *control;
         std::set<const IR::Parameter *> toDereference;
         std::vector<cstring> saveAction;
         P4::P4CoreLibrary &p4lib;
 
-        explicit UBPFControlBodyTranslator(const UBPFControl *control);
```
Comment (osinstom) [inline backends/ubpf/ubpfControl.h]: Why we had to remove `const` initializer?

File: backends/ubpf/ubpfControl.h [hunk 6]
```diff
@@ -69,6 +77,10 @@ namespace UBPF {
 
         std::set<const IR::Parameter *> toDereference;
         std::map<cstring, UBPFTable *> tables;
+        std::map<cstring, UBPFRegister *> registers;
+        std::vector<UBPFRegister *> registersLookups;
+        std::vector<cstring> registerKeys;
+        std::vector<cstring> pointerVariables;
```
Comment (osinstom) [inline backends/ubpf/ubpfControl.h]: Should `registersLookups`, `registersKeys` and `pointerVariables` be the members of `UBPFControl` class? Perhaps we can move them to `UBPFControlBodyTranslator`?

File: backends/ubpf/ubpfModel.h [hunk 7]
```diff
@@ -40,14 +40,29 @@ namespace UBPF {
...
+                           initial_value("initial_value"),
+                           index("index"),
+                           value("value") {}
+        ::Model::Elem sizeParam;
+        ::Model::Elem read;
+        ::Model::Elem write;
+
+        ::Model::Elem initial_value;
+        ::Model::Elem index;
+        ::Model::Elem value;
```
Comment (osinstom) [inline backends/ubpf/ubpfModel.h]: I think those variables should be moved before constructor.

File: backends/ubpf/examples/rate-limiter-structs.p4 (hunk 1 above)
Comment (kmateuszssak) [reply backends/ubpf/examples/rate-limiter-structs.p4]: They are useless. I am removing them.

File: backends/ubpf/ubpfModel.h (hunk 7 above)
Comment (osinstom) [reply backends/ubpf/ubpfModel.h]: We should also define access permissions for constructor and variables (see `UBPFModel`). The constructor is `protected` while variables are `public`.

File: backends/ubpf/examples/rate-limiter-structs.p4 (hunk 1 above)
Comment (kmateuszssak) [reply backends/ubpf/examples/rate-limiter-structs.p4]: Done.

File: backends/ubpf/ubpfTable.cpp [hunk 8]
```diff
@@ -16,69 +16,369 @@ limitations under the License.
...
                 auto ef = mi->to<P4::ExternFunction>();
                 if (ef != nullptr) {
                     if (ef->method->name.name == program->model.drop.name) {
-                        builder->emitIndent();
                         builder->append("pass = false");
                         return false;
+                    } else if (ef->method->name.name ==
+                               program->model.get_timestamp_ns.name) {
+                        builder->emitIndent();
+                        builder->append("ubpf_time_get_ns()");
```
Comment (osinstom) [inline backends/ubpf/ubpfTable.cpp]: I think `ubpf_time_get_ns()` should be global constant value.

File: backends/ubpf/ubpfTable.cpp [hunk 9]
```diff
@@ -534,5 +728,158 @@ namespace UBPF {
...
+                    const_value_counter == 0 ? "const_value" : "const_value_" +
+                                                               std::to_string(
+                                                                       const_value_counter -
+                                                                       1);
+        }
+
+        const_value_counter++;
+        target->emitTableLookup(codeGen, builder, dataMapName, keyName);
+    }
+
```
Comment (osinstom) [inline backends/ubpf/ubpfTable.cpp]: Maybe we should create a new file `ubpfRegister.cpp` and move those functions there?

File: backends/ubpf/ubpfTable.h [hunk 10]
```diff
@@ -45,6 +50,8 @@ namespace UBPF {
             keyTypeName = program->refMap->newName(instanceName + "_key");
             valueTypeName = program->refMap->newName(instanceName + "_value");
             dataMapName = instanceName;
+            //TODO: determine from where take this value for simple table
+            size = 1024;
```
Comment (osinstom) [inline backends/ubpf/ubpfTable.h]: Some TODO left. Is it default value for register's size?

File: backends/ubpf/ubpfTable.h [hunk 11]
```diff
@@ -75,8 +82,24 @@ namespace UBPF {
         void emitAction(EBPF::CodeBuilder *builder, cstring valueName);
 
         void emitInitializer(EBPF::CodeBuilder *builder);
-
-        void emitTableDefinition(EBPF::CodeBuilder *pBuilder);
     };
 
+    class UBPFRegister final: public UBPFTableBase {
+    public:
+        static int const_value_counter;
```
Comment (osinstom) [inline backends/ubpf/ubpfTable.h]: This `const_value_counter` really need to be public variable? I would prefer to define it locally in the .cpp file (if possible).

File: backends/ubpf/target.cpp (hunk 2 above)
Comment (kmateuszssak) [reply backends/ubpf/target.cpp]: We have to have only two emitTableLookup. I am removing one emitTableUpdate.

Comment (kmateuszssak) [reply backends/ubpf/target.cpp]: Done.

File: backends/ubpf/ubpfControl.cpp (hunk 3 above)
Comment (kmateuszssak) [reply backends/ubpf/ubpfControl.cpp]: findStatementWhereVariableIsNotUsedAsPointer ?

File: backends/ubpf/ubpfControl.cpp (hunk 4 above)
Comment (kmateuszssak) [reply backends/ubpf/ubpfControl.cpp]: Done.

File: backends/ubpf/ubpfControl.h (hunk 5 above)
Comment (kmateuszssak) [reply backends/ubpf/ubpfControl.h]: After moving `registersLookups`, `registersKeys` and `pointerVariables` to `UBPFControlBodyTranslator` i did bring back `const`

File: backends/ubpf/ubpfControl.h (hunk 6 above)
Comment (kmateuszssak) [reply backends/ubpf/ubpfControl.h]: Yes we can. Done.

File: backends/ubpf/ubpfTable.h (hunk 11 above)
Comment (kmateuszssak) [reply backends/ubpf/ubpfTable.h]: Done.

File: backends/ubpf/ubpfTable.cpp (hunk 9 above)
Comment (kmateuszssak) [reply backends/ubpf/ubpfTable.cpp]: Done.

File: backends/ubpf/ubpfTable.cpp (hunk 8 above)
Comment (kmateuszssak) [reply backends/ubpf/ubpfTable.cpp]: Done.

**Pull request merged**